Support --net=container:<NameOrID> for windows (Shared Pod Network)

madhanrm opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/docker/docker/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
Enabled --net=container option for windows
Vendor in hcsshim changes
**- How I did it**
Support is already available in docker. Just enabled it for Windows.
**- How to verify it**
Verified the running the relevant cmdlets
**- Description for the changelog**
Support for Windows containers to share network Endpoints
Enable --net=container: for windows

This is a basic requirement for kubenetes

**- A picture of a cute animal (not mandatory but encouraged)**

madhanrm commented:
@jhowardmsft, @jmesser81, @msabansal - Please review changes

jhowardmsft commented:
Changes look OK, but there's a couple of things. Can you split into two commits - one for the revendor and one for the engine change. Also, your commit message needs updating - it's currently `Initial changes for supporting --net:container:<containername/id> for…` with two signed-off-by lines.

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "ns1" git@github.com:madhanrm/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842353430920
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

madhanrm commented:
@jhowardmsft  Done

vdemeester commented on runconfig/hostconfig_windows.go:
```diff
@@ -26,9 +26,41 @@ func ValidateNetMode(c *container.Config, hc *container.HostConfig) error {
+	if len(parts) > 2 {
 		return fmt.Errorf("invalid --net: %s", hc.NetworkMode)
 	}
+	if hc.NetworkMode.IsContainer() {
```
All those checks are common with the `unix` side. We should re-use them somehow instead of duplicating (it makes it harder to maintain if we duplicate).

madhanrm replied:
Done

madhanrm reviewed (commented):
Update the PR with the changes requested.

madhanrm reviewed (commented):
Updated the PR with the changes requested

vdemeester reviewed (approved):
LGTM 🦁
@madhanrm could you squash your commit (leaving 2 commits : one for the vendoring changes, one with the other changes)

/cc @johnstep @jhowardmsft

madhanrm commented:
@vdemeester Done

madhanrm commented:
Ping @johnstep @jhowardmsft

jhowardmsft commented:
LGTM. Ping @mavenugo for a quick once over too before merging.

LK4D4 commented:
ping @mavenugo

jhowardmsft commented:
The vendor commit can be removed - HCSShim as already been bumped to .12 on master

madhanrm commented:
@jhowardmsft  Done

madhanrm commented:
Ping @mavenugo

dineshgovindasamy commented on daemon/start_windows.go:
```diff
@@ -157,14 +157,17 @@ func (daemon *Daemon) getLibcontainerdCreateOptions(container *container.Contain
+		networkSharedContainerID = container.NetworkSharedContainerID
+	}
+	createOptions = append(createOptions, &libcontainerd.NetworkEndpointsOption{
+		Endpoints:                epList,
```
Do we need to add a version check here? This will only work with corresponding HCSShip API right?

madhanrm replied:
The corresponding HCSShim is already vendored in  and part of docker executable. The version will not have any impact there. There would be a version difference between docker and windows, in which case, the new param would have empty value and would be ignored or not used by old windows version

mavenugo replied:
@madhanrm okay. in that case, how will this behave with an older windows OS ? This should ideally fail in that case isnt it ?

madhanrm replied:
Currently it would be a No-Op, if we try this command on old windows version.

madhanrm replied:
@mavenugo I'll explore some options for adding a version check and throw error in docker for unsupported windows version

madhanrm replied:
@mavenugo - I see that we can add version check in hcsshim, and return meaningful error from there.

mavenugo commented:
@madhanrm can you pls point us to the corresponding HCSShim changes that brought in this support ?

madhanrm commented:
https://github.com/docker/docker/commit/9983a7b88f24b4cebdda1cfb1d5dd20fdd50820f

madhanrm commented:
Added a change to throw error in docker when --net:container is used with --isolation=hyperv

Commit:
(*) Support --net:container:<containername/id> for windows
(*) (vdemeester) Removed duplicate code across Windows and Unix wrt Net:Containers
(*) Return unsupported error for network sharing for hyperv isolation containers

Signed-off-by: Madhan Raj Mookkandy <MadhanRaj.Mookkandy@microsoft.com>

LK4D4 commented:
Going to merge it. Feel free to object in follow-ups.
